components/layout/display_list_builder.rs

```diff
@@ -322,6 +329,47 @@ impl FragmentDisplayListBuilding for Fragment {
         }
     }
 
+    fn compute_background_image_size(&self,
+                                     style: &ComputedValues,
+                                     bounds: &Rect<Au>,
+                                     image: &png::Image)
+                                     -> Size2D<Au> {
+        // If `aspect_ratio` < 1.0, the image is tall; otherwise, it is wide.
+        let aspect_ratio = (image.width as f64) / (image.height as f64);
+        let intrinsic_size = Size2D(Au::from_px(image.width as int),
+                                    Au::from_px(image.height as int));
+        match (style.get_background().background_size.clone(), aspect_ratio < 1.0) {
+            (background_size::T::Contain, false) | (background_size::T::Cover, true) => {
+                Size2D(bounds.size.width,
+                       Au::from_frac_px(bounds.size.width.to_subpx() / aspect_ratio))
+            }
+
+            (background_size::T::Contain, true) | (background_size::T::Cover, false) => {
+                Size2D(Au::from_frac_px(bounds.size.height.to_subpx() * aspect_ratio),
+                       bounds.size.height)
+            }
+
+            (background_size::T::Explicit(background_size::ExplicitSize {
+                width,
+                height: LengthOrPercentageOrAuto::Auto,
+            }), _) => {
+                let width = MaybeAuto::from_style(width, bounds.size.width)
+                                      .specified_or_default(intrinsic_size.width);
+                Size2D(width, Au::from_frac_px(width.to_subpx() / aspect_ratio))
+            }
+
```

Comment:
Still seems to be an issue: A case seems to be missing for auto specified width and non-auto specified height of the background image.